### ISSUE TITLE
Update TypeScript type definitions

### ISSUE THREAD

**Issue by Fydon** [keycloak/keycloak-nodejs-connect#301]
### Description

The TypeScript type definitions are out of sync with the JavaScript code which can make it difficult to use, especially for new users. Ideally these should be updated to be in sync with JavaScript and any contributions should require updating these.

This was previously raised in Jira: [KEYCLOAK-13841](https://issues.redhat.com/browse/KEYCLOAK-13841) and [KEYCLOAK-14578](https://issues.redhat.com/browse/KEYCLOAK-14578).

### Motivation

TypeScript is very popular and the community has been trying to update these type definitions (https://github.com/keycloak/keycloak-nodejs-connect/pull/245 and https://github.com/keycloak/keycloak-nodejs-connect/pull/250), but these pull requests have yet to be merged.

**Comment by abstractj** [comment]
@stianst @Fydon [long time ago we discussed the idea of supporting Typescript](https://github.com/keycloak/keycloak-nodejs-connect/pull/123), but I'm no expert on this. I'm supportive to have Typescript support for the Node.js adapter, as long as we have enough people to review. That means getting some help from @ssilvert @jonkoops @edewit @agagancarczyk

**Comment by jonkoops** [comment]
Yeah, it's an issue that these typings are no longer up-to-date. I think we're due for re-writing the source code in TypeScript so that this cannot happen in the future, but this might be something that we need to factor into a larger effort to re-write the code behind our adapters.

@abstractj I have no problem diving in to check out the work that needs to be done for this, but for the time being I am occupied with my work for the new admin ui.

**Comment by abstractj** [comment]
@jonkoops the only work necessary for now is related to review PRs with typescript definitions. If that's ok, next year I will request your team's input on some of them.

**Comment by edewit** [comment]
I'm up to the task to do some reviews on stuff like that

**Comment by abstractj** [comment]
@Fydon if you would like to contribute and update the typescript definitions, you are more than welcome. We should now have reviewers for this.

**Comment by nbolten** [comment]
I would also be ready to translate the whole library to TypeScript-first, generating a JS library + .d.ts for distribution. But I want to check whether there would be reviewers and timely acceptance of a good pull request before doing that translation (since there are at least four TS type definition pull requests with no action).

**Comment by jonkoops** [comment]
@nbolten I agree that TypeScript source would be better and I would welcome any efforts in that respect. I am going to do some work to see what it would take to get it into the project.

**Comment by helio-frota** [comment]
I have a little xp in TS atm, and I'm not part of the KC team, in other words (just a personal opinion here), but when looking at the overall situation:

* keycloak-nodejs-admin-client - `TS`
* keycloak-admin-ui - `TS`
* keycloak-nodejs-connect - `JS`

IMO, moving to TS is more healthy for the KC team and contributors.

**Comment by jonkoops** [comment]
Yeah, we need a strategy on how to adopt it incrementally as I don't believe in a complete re-write just to accommodate the types. I am working on small quality of life stuff first, after which we can start swapping out tooling such linters and compilers with something more TypeScript friendly.

After that I think it makes sense to convert the code in small chunks from TS to JS.

**Comment by Fydon** [comment]
I assume that this is the NodeJS library that is [now deprecated](https://www.keycloak.org/2022/02/adapter-deprecation)? Any discussion about the deprecation should probably be made in [discussions](https://github.com/keycloak/keycloak/discussions/9818), but I thought I'd mention this here in case anyone was trying to update the Typescript definitions.
